rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala

```diff
@@ -257,12 +257,98 @@ object ProcNormalizeMatcher {
+                  (result.chan :: acc._1, result.knownFree)
+                }
+              )
+              (formalsResults._1.reverse, chan, formalsResults._2)
+            }
+          }
+        val (bindings, persistent) = p.receipt_ match {
+          case rl: ReceiptLinear =>
+            rl.receiptlinearimpl_ match {
+              case ls: LinearSimple =>
+                (ls.listlinearbind_.asScala.toList.map {
+                  case lbi: LinearBindImpl => (lbi.listname_.asScala.toList, lbi.name_)
+                  case _                   => throw new Error("Unexpected LinearBind production.")
+                }, false)
+              case _ => throw new Error("Unexpected LinearReceipt production.")
+            }
+          case rl: ReceiptRepeated =>
+            rl.receiptrepeatedimpl_ match {
+              case ls: RepeatedSimple =>
+                (ls.listrepeatedbind_.asScala.toList.map {
+                  case lbi: RepeatedBindImpl => (lbi.listname_.asScala.toList, lbi.name_)
+                  case _                     => throw new Error("Unexpected RepeatedBind production.")
+                }, true)
+              case _ => throw new Error("Unexpected RepeatedReceipt production.")
+            }
+        }
+        val (sources, thisLevelFree) = processSources(bindings)
+        val receipts                 = ReceiveSortMatcher.preSortBinds(processBindings(sources))
+        val groupedFrees = (DebruijnLevelMap[VarSort]() /: receipts)((env, receipt) =>
+          env.absorbFree(receipt._3) match {
+            case (newEnv, Nil) => newEnv
+            case _ =>
+              throw new Error("Free variable used as binder may not be used twice.")
+        })
+        val freeCount  = groupedFrees.next
+        val binds      = receipts.map((receipt) => (receipt._1, receipt._2))
+        val updatedEnv = input.env.absorbFree(groupedFrees)._1
+        val bodyResult =
+          normalizeMatch(p.proc_, ProcVisitInputs(Par(), updatedEnv, thisLevelFree))
+        ProcVisitOutputs(input.par.prepend(Receive(binds, bodyResult.par, persistent, freeCount)),
+                         bodyResult.knownFree)
+      }
+
       case p: PPar => {
         val result       = normalizeMatch(p.proc_1, input)
         val chainedInput = input.copy(knownFree = result.knownFree, par = result.par)
         normalizeMatch(p.proc_2, chainedInput)
       }
 
+      case p: PNew => {
+        import scala.collection.JavaConverters._
+        // TODO: bindings within a single new shouldn't have overlapping names.
+        val newBindings = p.listnamedecl_.asScala.toList.map {
+          case n: NameDeclSimpl => (Some(n.var_), NameSort)
+        }
+        val newEnv     = input.env.newBindings(newBindings)._1
+        val newCount   = newEnv.next - input.env.next
+        val bodyResult = normalizeMatch(p.proc_, ProcVisitInputs(Par(), newEnv, input.knownFree))
+        val foldedNew = bodyResult.par.singleNew() match {
```

Comment:
This is some neat recursion here.

Reply:
It's not recursion.

We're flattening new x in { new y in { P } } to new x, y in { P } by checking if the result of the body is a single new.

Reply:
It is recursion in the sense that you're calling into the function again. Or I don't actually understand the code.

Reply:
And by function I meant the case

case p: PNew

in case you have new x in { new y in { ...